bosh-dns recursor retry

HappyTobi opened the pull request:
Hi all,

I just implemented the retry mechanism for the `ServeDNS` requests.
The matching issue for that pr was #74
The issue was that it could happen that the  [`client.Exchange`](https://github.com/HappyTobi/bosh-dns-release/blob/0c0df677751e33b6dc21ba1df4c1e70f28bab2d5/src/bosh-dns/dns/server/handlers/forward_handler.go#L62) call gets en error because of an timeout or any other related network issue. 

So with the new implementation it's possible to setup and configure a `retry_count` with the new `recursor_retry_count` property at the bosh-dns configuration.
With that `retry` a failed response will be triggered again till the counter was reached. 

@mrosecrance
What do you think about the implementation? Can you give me some feedback?


I've also tested the new bosh-dns build on a running instance and the log of the fix will look like:
```
[FailoverRecursor] 2021-01-28T17:26:48.296952065Z DEBUG - dns request error received NXDOMAIN for foo.bar.xyz. from upstream (recursor: XXX.XX.XXX.XX:53) no retry - XXX.XX.XXX.XX:53
```

and for a network / i/o timeout:
```
[ForwardHandler] 2021-01-28T17:26:541327202Z ERROR - error recursing for tacoda.net. to "XXX.XX.XXX.XX:53": read udp XX.X.X.X:47601->XXX.XX.XXX.XX: i/o timeout
[FailoverRecursor] 2021-01-28T17:26:541327202Z ERROR - dns request network error read udp XX.X.X.X:47601->XXX.XX.XXX.XX:53: i/o timeout retry [1/3] for recoursor XXX.XX.XXX.XX53
[ForwardHandler] 2021-01-28T17:26:53.541652001Z ERROR - error recursing for tacoda.net. to "XXX.XX.XXX.XX:53": read udp XX.X.X.X:47601->XXX.XX.XXX.XX: i/o timeout
[FailoverRecursor] 2021-01-28T17:26:53.541687501Z ERROR - dns request network error read udp XX.X.X.X:47601->XXX.XX.XXX.XX:53: i/o timeout retry [2/3] for recoursor XXX.XX.XXX.XX53
[ForwardHandler] 2021-01-28T17:26:542092900Z ERROR - error recursing for tacoda.net. to "XXX.XX.XXX.XX:53": read udp XX.X.X.X:47601->XXX.XX.XXX.XX: i/o timeout
[FailoverRecursor] 2021-01-28T17:26:542092900Z ERROR - dns request network error read udp XX.X.X.X:47601->XXX.XX.XXX.XX:53: i/o timeout retry [3/3] for recoursor XXX.XX.XXX.XX53
```

HappyTobi committed:
NEW add configuration for recursor retry count

HappyTobi committed:
NEW add retry logic to smart recursor

HappyTobi committed:
UPDATE handle only network issues

HappyTobi committed:
UPDATE add error log to network error response

beyhan commented on src/bosh-dns/dns/server/handlers/forward_handler.go:
```diff
@@ -108,10 +108,19 @@ func (ForwardHandler) network(responseWriter dns.ResponseWriter) string {
+
+	switch err.(type) {
+	case net.Error:
+		r.logger.Error(r.logTag, fmt.Sprintf("received a network error from server rcode=%d - %s", dns.RcodeServerFailure, err.Error()))
```
I think this fits better at the end of the `performRetryLogic` function as a summery, why the retry wasn't successful. There we will cover automatically also other types of errors in case we decide to extend the list.

HappyTobi replied:
Fine for me, I'll move it to the retry logic

beyhan commented on src/bosh-dns/dns/server/handlers/failover_recursor_pool.go:
```diff
@@ -89,26 +111,48 @@ func newSmartFailoverRecursorPool(recursors []string, logger logger.Logger) Recu
 }
 
+func performRetryLogic(work func(string) error, recursor string, retryCount int, logTag string, log logger.Logger) (err error) {
+	if retryCount == 0 {
+		err = work(recursor)
+		return err
+	}
+	for ret := 0; ret < retryCount; ret++ {
+		err = work(recursor)
+		if err == nil {
+			return err
+		}
+		if _, ok := err.(net.Error); !ok {
+			log.Debug(logTag, fmt.Sprintf("dns request error %v no retry - %v\n", err, recursor))
+			return err
+		}
+
+		log.Error(logTag, fmt.Sprintf("dns request network error %s retry [%d/%d] for recoursor %s \n", err.(net.Error), ret+1, retryCount, recursor))
+	}
```
I think we are missing one execution in the retry loop, because e.g. it will be executed only once for ` retryCount=1`.  That means we will execute only one DNS resolution and will not have two DNS resolutions (the normal one + one retry). Or am I missing something here? 

I think it makes more sense to have the firs if statement `if retryCount == 0`  integrated in the `retry` loop , so that the retry loop checks something like `ret < retryCount + 1`. In this way we will have also a fix for the comment above. It would be also nice to have a unit test for the retry logic. WDYT?

HappyTobi replied:
Hi @beyhan,

sorry for the delay.
Thx for the feedback I will check that and update the code.

There was already a Test for the retry logic you have to check new forward_handler_test file.

beyhan commented on src/bosh-dns/dns/server/handlers/failover_recursor_pool.go:
```diff
@@ -89,26 +111,48 @@ func newSmartFailoverRecursorPool(recursors []string, logger logger.Logger) Recu
 	return ErrNoRecursorResponse
 }
 
+func performRetryLogic(work func(string) error, recursor string, retryCount int, logTag string, log logger.Logger) (err error) {
```
Naming is really hard and I'm not sure what fits better but `performRetryLogic` leaves the impression in me that the function does only retry logic. In fact the function does the initial DNS resolution and retry logic. What do you think about a name like `performWithRetryLogic` ?

HappyTobi replied:
Fine for me

HappyTobi committed:
UPDATE improve logging and cleanup code

beyhan commented:
@HappyTobi thank you for the update. I reviewed the latest version and it looks good for me.

FlorianNachtigall committed:
Make recursor retries configurable

This commit makes it configurable with the release property spec section.

Co-authored-by: joergdw <joerg.weisbarth@sap.com>

FlorianNachtigall commented:
In order to make the `recursor_retry_count` property configurable from the outside, we created a PR including it in the bosh-dns job spec section of the release: https://github.com/HappyTobi/bosh-dns-release/pull/1

@HappyTobi if you agree, feel free to merge it and thereby include it in this PR

HappyTobi commented:
Hi @FlorianNachtigall I will check that, and give you some feedback if needed, else I will merge it into my pr.

mrosecrance commented on src/bosh-dns/dns/server/handlers/factory.go:
```diff
@@ -49,7 +49,7 @@ func (f *Factory) CreateForwardHandler(recursors []string, cache bool) dns.Handl
 	// The default behavior defined by DNS spec is to use
 	// "smart" recursor selection.
-	pool := NewFailoverRecursorPool(f.shuffler.Shuffle(recursors), config.SmartRecursorSelection, f.logger)
+	pool := NewFailoverRecursorPool(f.shuffler.Shuffle(recursors), config.SmartRecursorSelection, 0, f.logger)
```
Is there a reason why you aren't using the retry config property here as well?

HappyTobi replied:
Nope, I change that to use the configuration parameter sorry for that

mrosecrance commented on src/bosh-dns/dns/config/config.go:
```diff
@@ -22,6 +22,7 @@ type Config struct {
 	Address            string       `json:"address"`
 	Port               int          `json:"port"`
 	BindTimeout        DurationJSON `json:"timeout,omitempty"`
+	RecursorRetryCount int          `json:"recursor_retry_count,omitempty"`
```
I don't think you've added this property to the job spec yet

HappyTobi replied:
Yes, I will merge the PR from @FlorianNachtigall to get that. (He implemented all the stuff at the spec)

mrosecrance commented on src/bosh-dns/dns/server/handlers/failover_recursor_pool.go:
```diff
@@ -89,26 +112,45 @@ func newSmartFailoverRecursorPool(recursors []string, logger logger.Logger) Recu
+			return err
+		}
+		if _, ok := err.(net.Error); !ok {
+			log.Debug(logTag, fmt.Sprintf("dns request error %v no retry - %v\n", err, recursor))
```
I think this is basically already logged when we switch recursors?

HappyTobi replied:
You log it at the the worker function:
https://github.com/cloudfoundry/bosh-dns-release/blob/1e5250dd8bb790aede2a6c3d2f3999c573fc5d2d/src/bosh-dns/dns/server/handlers/forward_handler.go#L66

So I think it's fine to delete that line that we don't log to much

mrosecrance commented on src/bosh-dns/dns/server/handlers/failover_recursor_pool.go:
```diff
@@ -89,26 +112,45 @@ func newSmartFailoverRecursorPool(recursors []string, logger logger.Logger) Recu
+			log.Debug(logTag, fmt.Sprintf("dns request error %v no retry - %v\n", err, recursor))
+			return err
+		}
+		log.Error(logTag, fmt.Sprintf("dns request network error %s retry [%d/%d] for recoursor %s \n", err.(net.Error), ret+1, retryCount, recursor))
```
Can we switch this to debug? Given that this feature is to optimize packet drop behavior, this doesn't seem like an error worth especially calling out.

HappyTobi replied:
Yes, fine for me

HappyTobi replied:
> We're deviating from my understanding of [this doc](https://tools.ietf.org/html/rfc1536) now with a retry on the same recursor. I think that's probably fine as one advantage I can see is that it will help call out down recursors vs just recursors that don't have the answer.

Yes, there is no part at the **rfc** that matches 100% so we have to go some small things different here.

Think the implementation was fine at this point, because we are using UDP and have a **fast retransmission** for the lost packages (thats the most important thing).

The retry is also not the "default" way... so in most of the cases the retry will never be called so we don't loose any time here and didn't add to much complexity.

friegger replied:
It would be undiscoverable how many retries happened unless debug logging is turned on, which we would not do in production.

mrosecrance replied:
Would info be more appropriate then? I tend to think of errors as indicators of a failing system and since we're building in some resiliency to that, I think the error below is the real indicator. 

Also in the 0 retry default case, we'd get two errors for the same packet loss which could be confusing?

mrosecrance reviewed (commented):
We're deviating from my understanding of [this doc](https://tools.ietf.org/html/rfc1536) now with a retry on the same recursor. I think that's probably fine as one advantage I can see is that it will help call out down recursors vs just recursors that don't have the answer.

HappyTobi commented:
> We're deviating from my understanding of [this doc](https://tools.ietf.org/html/rfc1536) now with a retry on the same recursor. I think that's probably fine as one advantage I can see is that it will help call out down recursors vs just recursors that don't have the answer.

Yes, there is no part at the rfc that matches 100% so we have to go some small things different here.

Think the implementation was fine at this point, because we are using UDP and have a fast retransmission for the lost packages (thats the most important thing).

The retry is also not the "default" way... so in most of the cases the retry will never be called so we don't loose any time here and didn't add to much complexity.

HappyTobi committed:
UPDATE add recursor to handler factory, change logging

FlorianNachtigall committed:
Rename bosh-dns property recursor_retry_count to recursor_max_retries

mrosecrance commented:
Sam and I looked at the changes and it looks good. Currently the pipeline is broken due to a go 1.16 bump so merging is blocked on that being fixed.

friegger commented:
We were wondering wether `recursor_retry_count` is giving away the right meaning or whether `recursor_max_retries` would be more appropriate?

jpalermo commented:
Pipeline is fixed and we can merge this now. Do you want us to wait for a change to `recursor_retry_count`? Or is it ready to go as it is?

HappyTobi commented:
I will rename it and push it again.

FlorianNachtigall commented:
@HappyTobi We already renamed the property for our internal fork. I opened a PR with the respective changes: https://github.com/HappyTobi/bosh-dns-release/pull/2/

However, we were discussing whether it makes more sense to wait with the merge until we have collected some experience in production and round off the feature with our learnings.

To give some examples what those could be: Until now we've realized that it could be helpful to include the domain name or some alias uuid in the logs to identify and relate multiple dns resolution retries. Further, our operators found "retry [0/2]" a bit confusing and would have preferred something more like "try [1/3]".

What do you think @jpalermo & @mrosecrance?

HappyTobi committed:
FIX typo and retry count output

HappyTobi committed:
UPDATE retry logging

HappyTobi commented:
Hi @jpalermo & @mrosecrance & @FlorianNachtigall 

from my side it's fine now.

I fixed some typos and changed the `retry` output complete because to  switch from  `[0/2]` to `[1/3]`  in the output also doesn't represent whats really happen because the user set  `max_retry_count` to `2` and see something like `[1/3]` in the output .

The new output for `max_retry_count = 2` looks like: 
```
dns request network error lookup : i/o timeout retry [1/1] - request count [2] for recursor retry
```

Regards

mrosecrance commented:
I think the last open thread on this is whether `log.Debug(logTag, fmt.Sprintf("dns request error %v no retry - %v\n", err, recursor))` should be switched to INFO or WARN but I don't have a strong opinion on that.
